crates/executor/src/substate.rs

```diff
@@ -0,0 +1,289 @@
+                        let val = entry.get_mut();
+                        if val.decr_storage_counter() {
+                            // remove account that we didn't change from substate
+
+                            entry.remove_entry();
+                            return
+                        }
+                        val.info = Account::default();
+                        val.storage.clear();
+                    }
+                };
+            }
+            AccountInfoChangeSet::Destroyed { old } => match self.accounts.entry(*address) {
+                Entry::Vacant(_) => {
+                    // We inserted this account in apply fn.
+                    panic!("It should be present, something is broken");
+                }
+                Entry::Occupied(mut entry) => {
+                    let val = entry.get_mut();
+
+                    // Contrary to Created we are not removing this account as we dont know if
+                    // this account was changer or not by `Changed` changeset.
+                    val.decr_storage_counter();
+                    val.info = *old;
+                }
+            },
+            AccountInfoChangeSet::Changed { old, .. } => {
+                self.accounts.entry(*address).or_default().info = *old;
+            }
+            AccountInfoChangeSet::NoChange => (),
+        }
+    }
+}
+/// Account changes in substate
+#[derive(Debug, Clone, Default)]
+pub struct AccountSubState {
+    /// New account state
+    pub info: Account,
+    /// If account is selfdestructed or newly created, storage will be cleared.
+    /// and we dont need to ask the provider for data.
+    /// As we need to have as
+    pub storage_is_clear: Option<u32>,
+    /// storage slots
+    pub storage: HashMap<H256, U256>,
+}
+
+impl AccountSubState {
+    /// Increment storage counter to mark this storage was cleared
+    pub fn inc_storage_counter(&mut self) {
+        self.storage_is_clear = Some(self.storage_is_clear.unwrap_or_default() + 1);
+    }
+
+    /// Decrement storage counter to represent that changeset that cleared storage was reverted.
+    pub fn decr_storage_counter(&mut self) -> bool {
+        let Some(cnt) = self.storage_is_clear else { return false};
+
+        if cnt == 1 {
+            self.storage_is_clear = None;
+            return true
+        }
+        false
+    }
+    /// Account is created
+    pub fn created_account(info: Account) -> Self {
+        Self { info, storage_is_clear: Some(1), storage: HashMap::new() }
+    }
+    /// Should we ask the provider for storage data
+    pub fn ask_provider(&self) -> bool {
+        self.storage_is_clear.is_none()
+    }
+}
+
+/// Wrapper around substate and provider, it decouples the database that can be Latest or historical
+/// with substate changes that happened previously.
+pub struct SubStateWithProvider<'a, PROVIDER: StateProvider> {
+    /// Substate
+    pub substate: &'a mut SubStateData,
```

Comment:
why are u taking the mut ref here?

Reply:
Will need to check, maybe it is just a leftover from the previous iteration